### PR TITLE
Extract data-fetching layer using TRPC

### PR DESCRIPTION
**Pull request by swh00tw**
## Description

1. Extract the data-fetching logic using [trpc](https://trpc.io/) to make future transition smoother.
2. Refactor Application layer to adopt new data models.
3. Change `dateSchema` in `recnet-api-model` from `z.date()` to `z.string().datetime()` since we decided to use ISO8601 date string to represent timestamp. (And there's no way for http response to transmit `Date` object. It's un-serializable lol. So I just fixed a bug.)
4. Remove old data schemas and old serverless function endpoints and server actions 🎉.

## Related Issue

#108 

### Briefly introduction of TRPC

Trpc is a tool to provide type-safe API fetching between client & server (In my case, it sits between client-side aka browser and server-side aka Next.js's backend serverless functions). 

On the server side, you defined trpc procedures (api endpoints) and trpc will turn them into API endpoints. It will be looked like this. 
```typescript
// in /server/routers/...
const router = t.router({
  getMe: publicProcedure
      .input(z.object({...})) // write the schema of what input should be passed to this endpoint
      .output(z.object({ user: userSchema.nullable() })) // write the schema of what this endpoint will output
      .query(async (opts) => {
        // write some http fetchs or db operations here
        const { tokens } = opts.ctx;
        if (!tokens) {
          return {
            user: null,
          };
        }
        const user = await getUserByTokens(tokens);
        return {
          user,
        };
      }),
   // other procedures...
})
```
It can be integrated will `zod` to do input/output validation, which is useful for catching wrong data shape.

On the client side, trpc provides a useful query client which inherits the popular data-fetching library [`tanstack-query`](https://tanstack.com/query/v5) (formerly called [`react-query`](https://tanstack.com/query/v5)), which offers rich functionalities such as caching, refetching, loading states, invalidation, and pagination queries. The most important part is trpc client has already wrapped the actual http fetch deep inside and offered the types automatically. (It makes the API calls just like function calls. Just like remote procedure call. That's why it is called trpc.)

So, instead of doing http fetch like this 
```typescript
const res = await fetch("https://api/some/endpoint/...")
// ^^ the type of "res" is any here, you need to do some schema validation or use 'as' to assert the type, which is ugly
```

You just use 
```typescript
const { data, isPending, isError, isFetching } = trpc.getMe.useQuery();
// ^^ the type of "data" here is { user: User | null } just like we defined in our server side code automatically
```

If you are interested in learning more, this might be a good entry point:
- https://trpc.io/docs/concepts 

### Why using it?

Anyway, the benefits of using the trpc are listed below
- type-safety API fetching
- extracting a data fetching layer. All data-fetching logics are put under `/server/routers` folder. (Before this change, the [route handlers](https://nextjs.org/docs/app/building-your-application/routing/route-handlers) are put under `app/api/` and need to add a `route.ts` for each endpoint, which is annoying. And the [server action](https://nextjs.org/docs/app/building-your-application/data-fetching/server-actions-and-mutations) are put under `/server` folder.)
- integrate with `React-query` 
- Also support server-side rendering (fetching data server-side using `serverClient`)

### Why this PR gonna make future transition smoother.

See the diagram below.
Screenshot 2024-03-21 at 12 40 52 AM

## TODO

- [x] Paste the testing link: https://rec-net-git-refactor-integrate-trpc-recnet-542617e7.vercel.app/
- [x] Clear `console.log` or `console.error` for debug usage

### CODE REVIEW

File: apps/recnet/src/app/[handle]/Profile.tsx [hunk 1]
```diff
@@ -2,27 +2,24 @@
 
 import { zodResolver } from "@hookform/resolvers/zod";
```
Comment (joannechen1223) [inline apps/recnet/src/app/[handle]/Profile.tsx:3]: [nit] why did you include `[]` to the directory name here?

**Review by joannechen1223 — APPROVED** [review]
I've tested on the provided testing link. I encountered once 500 Error once while I was trying to login. However, I am not able to reproduce. Just put this issue here for your information, otherwise LGTM.

File: apps/recnet/src/app/[handle]/Profile.tsx (hunk 1 above)
Comment (swh00tw) [reply apps/recnet/src/app/[handle]/Profile.tsx:3]: It marks this is a dynamic route: https://nextjs.org/docs/pages/building-your-application/routing/dynamic-routes

The request to `/yoav`, `/swh00tw`, `joanne`, ... will be direct to this route

File: apps/recnet/src/app/[handle]/Profile.tsx [hunk 2]
```diff
@@ -2,27 +2,24 @@
...
+import { Dialog, Button, Flex, Text, TextField } from "@radix-ui/themes";
+import { TRPCClientError } from "@trpc/client";
 import { useRouter } from "next/navigation";
 import { useState } from "react";
-import { useForm } from "react-hook-form";
+import { useForm, useFormState } from "react-hook-form";
 import { toast } from "sonner";
 import * as z from "zod";
 
+import { trpc } from "@recnet/recnet-web/app//_trpc/client";
```
[the remark this answers is not in the document]
Comment (swh00tw) [reply apps/recnet/src/app/[handle]/Profile.tsx]: fixed

**Pull request merged**